In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [2]:
# Data generated by me flailing at my keyboard

raw_data = pd.DataFrame([1,2,5,7,8,9,1,2,4,6,7,8,10,4,6,2,5,7,14,15,13,12,11,11,11,16,17,14,3,5], columns = ['Raw'])

In [3]:
raw_data

Raw
0     1
1     2
2     5
3     7
4     8
5     9
6     1
7     2
8     4
9     6
10    7
11    8
12   10
13    4
14    6
15    2
16    5
17    7
18   14
19   15
20   13
21   12
22   11
23   11
24   11
25   16
26   17
27   14
28    3
29    5

In [4]:
ss1 = StandardScaler()

allscale = pd.DataFrame(ss1.fit_transform(raw_data), index = raw_data.index, columns = ['all_ss1'])
allscale

all_ss1
0  -1.487297
1  -1.270700
2  -0.620910
3  -0.187717
4   0.028880
5   0.245476
6  -1.487297
7  -1.270700
8  -0.837507
9  -0.404314
10 -0.187717
11  0.028880
12  0.462073
13 -0.837507
14 -0.404314
15 -1.270700
16 -0.620910
17 -0.187717
18  1.328459
19  1.545056
20  1.111863
21  0.895266
22  0.678669
23  0.678669
24  0.678669
25  1.761652
26  1.978249
27  1.328459
28 -1.054104
29 -0.620910

In [5]:
unscaledresults = pd.DataFrame(ss1.inverse_transform(allscale), index = allscale.index, columns = ['all_ss1_invert'])
unscaledresults

all_ss1_invert
0              1.0
1              2.0
2              5.0
3              7.0
4              8.0
5              9.0
6              1.0
7              2.0
8              4.0
9              6.0
10             7.0
11             8.0
12            10.0
13             4.0
14             6.0
15             2.0
16             5.0
17             7.0
18            14.0
19            15.0
20            13.0
21            12.0
22            11.0
23            11.0
24            11.0
25            16.0
26            17.0
27            14.0
28             3.0
29             5.0

In [6]:
resultsdf = pd.concat([raw_data, allscale, unscaledresults], axis = 1)

resultsdf

Raw   all_ss1  all_ss1_invert
0     1 -1.487297             1.0
1     2 -1.270700             2.0
2     5 -0.620910             5.0
3     7 -0.187717             7.0
4     8  0.028880             8.0
5     9  0.245476             9.0
6     1 -1.487297             1.0
7     2 -1.270700             2.0
8     4 -0.837507             4.0
9     6 -0.404314             6.0
10    7 -0.187717             7.0
11    8  0.028880             8.0
12   10  0.462073            10.0
13    4 -0.837507             4.0
14    6 -0.404314             6.0
15    2 -1.270700             2.0
16    5 -0.620910             5.0
17    7 -0.187717             7.0
18   14  1.328459            14.0
19   15  1.545056            15.0
20   13  1.111863            13.0
21   12  0.895266            12.0
22   11  0.678669            11.0
23   11  0.678669            11.0
24   11  0.678669            11.0
25   16  1.761652            16.0
26   17  1.978249            17.0
27   14  1.328459            14.0
28    3 -1.054104             3.0
29    5 -0.620910             5.0

In [7]:
check = resultsdf.Raw - resultsdf.all_ss1_invert

#These should all be zero
check

0     0.0
1     0.0
2     0.0
3     0.0
4     0.0
5     0.0
6     0.0
7     0.0
8     0.0
9     0.0
10    0.0
11    0.0
12    0.0
13    0.0
14    0.0
15    0.0
16    0.0
17    0.0
18    0.0
19    0.0
20    0.0
21    0.0
22    0.0
23    0.0
24    0.0
25    0.0
26    0.0
27    0.0
28    0.0
29    0.0
dtype: float64

In [8]:
#hypothetical train test split

#setting random seed for repeatability
np.random.seed(14)
#This selects 5 random intergers between 0 and 29, which we can map to the index of our underlyingdata
test_data_index = np.random.choice(range(30), size=5, replace=False)

In [9]:
test_data_index

array([ 2, 18, 15,  3, 25])

In [10]:
#The index values for our test data willl just be our 5 chosen numbers
test_data = raw_data.iloc[test_data_index]
display(test_data)

# Index values for the training data will be 0-29 minus those numbers.

train_data_index = list(range(30))

for i in test_data_index:
    train_data_index.remove(i)
train_data = raw_data.iloc[train_data_index]

display(train_data)


Raw
2     5
18   14
15    2
3     7
25   16

Raw
0     1
1     2
4     8
5     9
6     1
7     2
8     4
9     6
10    7
11    8
12   10
13    4
14    6
16    5
17    7
19   15
20   13
21   12
22   11
23   11
24   11
26   17
27   14
28    3
29    5

In [11]:
display(raw_data.value_counts())
display(train_data.value_counts())
display(test_data.value_counts())

Raw
7      3
2      3
11     3
5      3
1      2
14     2
4      2
6      2
8      2
3      1
17     1
16     1
10     1
12     1
13     1
15     1
9      1
dtype: int64

Raw
11     3
8      2
7      2
6      2
5      2
4      2
2      2
1      2
17     1
15     1
14     1
13     1
12     1
10     1
9      1
3      1
dtype: int64

Raw
16     1
14     1
7      1
5      1
2      1
dtype: int64

Ooooookkay. Now. Let's try some transforms on the training data seperately. 

Remember, the test data represents data the model has not seen yet.

In [12]:
sstrain = StandardScaler()

trainscaleresults = pd.DataFrame(sstrain.fit_transform(train_data), index = train_data.index, columns = ['train_sstrain'])

trainscaleinvert = pd.DataFrame(sstrain.inverse_transform(trainscaleresults), index = trainscaleresults.index, columns = ['train_sstrain_invert'])

trainscaleresults

train_sstrain
0       -1.506646
1       -1.281100
4        0.072175
5        0.297720
6       -1.506646
7       -1.281100
8       -0.830009
9       -0.378917
10      -0.153371
11       0.072175
12       0.523266
13      -0.830009
14      -0.378917
16      -0.604463
17      -0.153371
19       1.650995
20       1.199904
21       0.974358
22       0.748812
23       0.748812
24       0.748812
26       2.102087
27       1.425449
28      -1.055554
29      -0.604463

In [13]:
resultsdf = pd.concat([resultsdf, trainscaleresults, trainscaleinvert], axis = 1)

In [14]:
resultsdf

Raw   all_ss1  all_ss1_invert  train_sstrain  train_sstrain_invert
0     1 -1.487297             1.0      -1.506646                   1.0
1     2 -1.270700             2.0      -1.281100                   2.0
2     5 -0.620910             5.0            NaN                   NaN
3     7 -0.187717             7.0            NaN                   NaN
4     8  0.028880             8.0       0.072175                   8.0
5     9  0.245476             9.0       0.297720                   9.0
6     1 -1.487297             1.0      -1.506646                   1.0
7     2 -1.270700             2.0      -1.281100                   2.0
8     4 -0.837507             4.0      -0.830009                   4.0
9     6 -0.404314             6.0      -0.378917                   6.0
10    7 -0.187717             7.0      -0.153371                   7.0
11    8  0.028880             8.0       0.072175                   8.0
12   10  0.462073            10.0       0.523266                  10.0
13    4 -0.837507             4.0      -0.830009                   4.0
14    6 -0.404314             6.0      -0.378917                   6.0
15    2 -1.270700             2.0            NaN                   NaN
16    5 -0.620910             5.0      -0.604463                   5.0
17    7 -0.187717             7.0      -0.153371                   7.0
18   14  1.328459            14.0            NaN                   NaN
19   15  1.545056            15.0       1.650995                  15.0
20   13  1.111863            13.0       1.199904                  13.0
21   12  0.895266            12.0       0.974358                  12.0
22   11  0.678669            11.0       0.748812                  11.0
23   11  0.678669            11.0       0.748812                  11.0
24   11  0.678669            11.0       0.748812                  11.0
25   16  1.761652            16.0            NaN                   NaN
26   17  1.978249            17.0       2.102087                  17.0
27   14  1.328459            14.0       1.425449                  14.0
28    3 -1.054104             3.0      -1.055554                   3.0
29    5 -0.620910             5.0      -0.604463                   5.0

In [15]:
check = resultsdf.Raw - resultsdf.train_sstrain_invert

check 

0     0.000000e+00
1     0.000000e+00
2              NaN
3              NaN
4     0.000000e+00
5     0.000000e+00
6     0.000000e+00
7     0.000000e+00
8     4.440892e-16
9     0.000000e+00
10    0.000000e+00
11    0.000000e+00
12    0.000000e+00
13    4.440892e-16
14    0.000000e+00
15             NaN
16    0.000000e+00
17    0.000000e+00
18             NaN
19    0.000000e+00
20    0.000000e+00
21    0.000000e+00
22    0.000000e+00
23    0.000000e+00
24    0.000000e+00
25             NaN
26    0.000000e+00
27    0.000000e+00
28    0.000000e+00
29    0.000000e+00
dtype: float64

So far so good. But! Note the difference in our all_ss1 and the train_sstrain.

In [16]:
check = resultsdf.all_ss1 - resultsdf.train_sstrain

check


0     0.019349
1     0.010400
2          NaN
3          NaN
4    -0.043295
5    -0.052244
6     0.019349
7     0.010400
8    -0.007498
9    -0.025397
10   -0.034346
11   -0.043295
12   -0.061193
13   -0.007498
14   -0.025397
15         NaN
16   -0.016448
17   -0.034346
18         NaN
19   -0.105939
20   -0.088041
21   -0.079092
22   -0.070143
23   -0.070143
24   -0.070143
25         NaN
26   -0.123838
27   -0.096990
28    0.001451
29   -0.016448
dtype: float64

But, you may say, if I apply all steps to the test data that I applied to the training data, I should be fine. Let's look at that w/r/t our scaler:

In [17]:
# Here I have copy and pasted all my code from above, and will simply put in 'test' everwhere there is a 'train'

# sstrain = StandardScaler()
sstest = StandardScaler()

# trainscaleresults = pd.DataFrame(sstrain.fit_transform(train_data), index = train_data.index, columns = ['train_sstrain'])
testscaleresults = pd.DataFrame(sstest.fit_transform(test_data), index = test_data.index, columns = ['test_sstest'])

# trainscaleinvert = pd.DataFrame(sstrain.inverse_transform(trainscaleresults), index = trainscaleresults.index, columns = ['train_sstrain_invert'])
testscaleinvert = pd.DataFrame(sstest.inverse_transform(testscaleresults), index = testscaleresults.index, columns = ['test_ssttest_invert'])

# trainscaleresults
testscaleresults

test_sstest
2     -0.711057
18     0.973026
15    -1.272418
3     -0.336817
25     1.347266

Looks fine and normal, but what if we tried to put this into a model fit on the training data?



In [18]:
# In the training data, the 'underlying' values associated with the values of test_sstest is: 
newxtest = sstrain.inverse_transform(testscaleresults)

newxtest

array([[ 4.52739368],
       [11.99409286],
       [ 2.03849395],
       [ 6.18666016],
       [13.65335934]])

In [19]:
# Compare that to the true values:

test_data

Raw
2     5
18   14
15    2
3     7
25   16

So if we fed this into our model, our predictions would be given out for the wrong data points. The model would be trying to give us the asnwer for an 'underlying' data point of approx 4.5, 12, 2, 6.2, 13.7 and so on. The predictions would be true to the model AT THOSE POINTS. But we'd be ascribing them to the underlying points of 5, 14, 2, 7, and 16.  This would, of course, generate inexplicably inaccurate results. 

Let's try to do this properly.

In [20]:
# Copy and pasted from above:

# sstrain = StandardScaler()
# sstest = StandardScaler()
# I'm commenting out this option entirely, we *don't* want to generate a new scaler 


# trainscaleresults = pd.DataFrame(sstrain.fit_transform(train_data), index = train_data.index, columns = ['train_sstrain'])
# testscaleresults = pd.DataFrame(sstest.fit_transform(test_data), index = test_data.index, columns = ['test_sstest'])
# Here we'll keep most things as above, but we will only transform using the training scaler. I'll also adjust the column name. 

testscaleresults = pd.DataFrame(sstrain.transform(test_data), index = test_data.index, columns = ['test_sstrain'])

# trainscaleinvert = pd.DataFrame(sstrain.inverse_transform(trainscaleresults), index = trainscaleresults.index, columns = ['train_sstrain_invert'])
testscaleinvert = pd.DataFrame(sstrain.inverse_transform(testscaleresults), index = testscaleresults.index, columns = ['test_ssttest_invert'])

# trainscaleresults
display(testscaleresults)
display(testscaleinvert)

test_sstrain
2      -0.604463
18      1.425449
15     -1.281100
3      -0.153371
25      1.876541

test_ssttest_invert
2                   5.0
18                 14.0
15                  2.0
3                   7.0
25                 16.0

Tadah!

### Okay, let's see this in action with a model.

For this dataset, let's assume we're a restaurant, catering exclusively to couples, and we want to calculate how many chairs we'll need on a night, given the number of couples who have booked a reservation. We also keep 10 tables (so 20 chairs) for walk in customers. 

I will generate the X data by randomly hitting my keyboard, and y data by taking the X, multiplying by 2, and adding 10.

In [21]:
x = [12,14,12,23,53,73,1,3,45,32,14,24,74,86,2,5,67,34,7,2,64,435,98,2,6,234,45,14,34,8]

In [22]:
len(x)

30

In [23]:
couplesandchairs_df = pd.DataFrame(x, columns = ['reservation_couples'])

In [24]:
couplesandchairs_df['chairs'] = couplesandchairs_df.reservation_couples*2 + 20

couplesandchairs_df

reservation_couples  chairs
0                    12      44
1                    14      48
2                    12      44
3                    23      66
4                    53     126
5                    73     166
6                     1      22
7                     3      26
8                    45     110
9                    32      84
10                   14      48
11                   24      68
12                   74     168
13                   86     192
14                    2      24
15                    5      30
16                   67     154
17                   34      88
18                    7      34
19                    2      24
20                   64     148
21                  435     890
22                   98     216
23                    2      24
24                    6      32
25                  234     488
26                   45     110
27                   14      48
28                   34      88
29                    8      36

In [25]:
X_train, X_test, y_train, y_test = train_test_split(couplesandchairs_df.reservation_couples, couplesandchairs_df.chairs,  random_state = 14)

In [26]:
display(X_train.head())
display(y_train.head())
display(X_test.head())
display(y_test.head())

9      32
5      73
21    435
28     34
17     34
Name: reservation_couples, dtype: int64

9      84
5     166
21    890
28     88
17     88
Name: chairs, dtype: int64

2      12
18      7
15      5
3      23
25    234
Name: reservation_couples, dtype: int64

2      44
18     34
15     30
3      66
25    488
Name: chairs, dtype: int64

In [27]:
# Now let's build a model

sschairs = StandardScaler()

train_chairs_scaled = pd.DataFrame(sschairs.fit_transform(np.asarray(X_train).reshape(-1, 1)), index = X_train.index)

model = LinearRegression()

model.fit(train_chairs_scaled, y_train)

LinearRegression()

In [28]:
train_chairs_scaled

0
9  -0.228852
5   0.233983
21  4.320481
28 -0.206275
17 -0.206275
14 -0.567512
26 -0.082099
19 -0.567512
4   0.008210
0  -0.454625
29 -0.499780
13  0.380736
8  -0.082099
16  0.166251
23 -0.567512
10 -0.432048
6  -0.578801
7  -0.556223
22  0.516200
12  0.245272
24 -0.522357
11 -0.319161

In [29]:
model.predict(np.asarray(X_test).reshape(-1, 1))

array([ 2250.57138288,  1364.72724607,  1010.38959135,  4199.42848385,
       41582.05105699,  2604.9090376 , 11463.35040564,  2604.9090376 ])

In [30]:
X_test_scaled = pd.DataFrame(sschairs.transform(np.asarray(X_test).reshape(-1, 1)), index = X_test.index)
X_test_scaled

0
2  -0.454625
18 -0.511069
15 -0.533646
3  -0.330450
25  2.051459
1  -0.432048
20  0.132385
27 -0.432048

In [31]:
y_hat = model.predict(X_test_scaled)
y_hat

array([ 44.,  34.,  30.,  66., 488.,  48., 148.,  48.])

In [32]:
y_test 

2      44
18     34
15     30
3      66
25    488
1      48
20    148
27     48
Name: chairs, dtype: int64

All well and good. Buuuuuut

In [33]:
X_test

2      12
18      7
15      5
3      23
25    234
1      14
20     64
27     14
Name: reservation_couples, dtype: int64

In [34]:
badscaler = StandardScaler()
badscaler.fit(np.asarray(X_test).reshape(-1, 1))

X_test_scaled_bad = pd.DataFrame(badscaler.transform(np.asarray(X_test).reshape(-1, 1)), index = X_test.index)

X_test_scaled_bad

0
2  -0.474494
18 -0.543013
15 -0.570420
3  -0.323752
25  2.567748
1  -0.447086
20  0.238103
27 -0.447086

In [35]:
X_test_scaled

0
2  -0.454625
18 -0.511069
15 -0.533646
3  -0.330450
25  2.051459
1  -0.432048
20  0.132385
27 -0.432048

In [36]:
y_hat_bad = model.predict(X_test_scaled_bad)
y_hat_bad

array([ 40.47993064,  28.34050481,  23.48473447,  67.18666748,
       579.47043769,  45.33570098, 166.72995932,  45.33570098])

In [37]:
y_test 

2      44
18     34
15     30
3      66
25    488
1      48
20    148
27     48
Name: chairs, dtype: int64